### PR TITLE
consvc-1670 - Enable deploying to production

### PR DESCRIPTION
**Pull request by Dexterp37**
This introduces the CircleCI facilities that are used to trigger deployment to production.

It also changes the release process documents to highlight the new release process.

Here's what the UI looks like (I tried this in my [toy repo](https://github.com/Dexterp37/test_ci)):

image

Works toward, but does not fully fix, #352 (it still lacks the SRE changes on the pipeline side).

### CODE REVIEW

File: docs/dev/release-process.md
Comment (ncloudioj) [inline docs/dev/release-process.md:31]: Just for clarification, are `developers` referring to the same individual(s) who authored and merged the pull request, or do we need someone else to operate this including the approval process mentioned below? I think it should be the former but just wanted to double check.

File: docs/dev/release-process.md
Comment (ncloudioj) [inline docs/dev/release-process.md:42]: Yep, don't panic!™ 😎

File: docs/dev/release-process.md
Comment (ncloudioj) [inline docs/dev/release-process.md:52]: Another clarification question: what if the problematic commit is not the last commit and some subsequent commits after it are also deployed in production. I think we'd want to provide developers with a way to easily revert to the last working commit. This can be tricky though with a highly active repo, but still worth thinking about.

**Review by ncloudioj — CHANGES_REQUESTED** [review]
The changes for CircleCI look good. I left a few comments for the documentation. I know those comments/asks are not easy to fully expand in here, so could you please provide a little more clarifications here, we can tackle those remaining details in a follow up perhaps in a different format (such as a CD runbook on mana). Thanks!

File: docs/dev/release-process.md
Comment (Dexterp37) [reply docs/dev/release-process.md:31]: I was intentionally generic in this statement, because anyone could trigger, as long as they have write access to the Merino repository.

Ideally speaking, I would expect the individual who authored the PR to do that, but we don't strictly enforce this. The reason why I would expect that is that authors know the code and know what to watch out for when things are hitting production. If the author touched search providers in a subtle way and the plots show anomalies with search queries, then they might be better suited to identify the problem.

In general, we want developers / authors to feel the _responsibility_ of deploying to production, own that deployment and shepherd it to conclusion.

File: docs/dev/release-process.md
Comment (Dexterp37) [reply docs/dev/release-process.md:44]: Yes, I believe this part should be changed / extended by somebody else rather than me. I meant to add this as an initial point of discussion, in order to have _something_. What I wrote is mostly high level and always relevant, regardless of the specifics of the incident and it boils down to this:

- if the problem is trivial and the fix is easy, create a PR, fix the thing and deploy;
- if the problem warrants a more thorough investigation, revert/turn off (if only we had feature flags) the latest change and investigate.

I'd recommend either leaving this text as it is or extending it in a separate PR as you see fit. I believe there's still value in slightly documenting the process to act on things when prod breaks.

File: docs/dev/release-process.md
Comment (Dexterp37) [reply docs/dev/release-process.md:52]: This is a great question, and I believe that this question boils down to the philosophical difference between Continuous Deployment and Continuous Delivery: with Continuous Delivery, you delivery whenever you feel like, maybe weekly, maybe not, but you potentially put together multiple changesets.

With Continuous Deployment, the idea is to deploy every time you merge to main. That's the part that reduces the risk of breaking, because when releasing on each single merge you can easily tell if you're merging something problematic. Ideally, you don't uncover with a deployment from today something that broke 15 days ago. Because the 15 days ago thing would have triggered problems on that deployment. If the interaction with the new thing is the problem, then you still solve it by reverting/disabling the newest change.

The need to go back in time to deploy an old commit, I believe, sort of defeats the point of Continuous Deployment. It's still doable, manually, by SRE though (which is the current state of things anyway).

File: .circleci/config.yml
Comment (hackebrot) [inline .circleci/config.yml:63]: TIL! 🤩

File: docs/dev/release-process.md
Comment (hackebrot) [reply docs/dev/release-process.md:31]: There's a lot of helpful context in your response, @Dexterp37. I would love this to be captured in the documentation.

File: .circleci/config.yml
Comment (hackebrot) [inline .circleci/config.yml:226]:
```suggestion
          name: Push to Docker Hub (prod)
```

File: docs/dev/release-process.md
Comment (hackebrot) [inline docs/dev/release-process.md:33]:
```suggestion
1. opening the [CircleCI dashboard][circleci_dashboard];
```

File: docs/dev/release-process.md
Comment (hackebrot) [reply docs/dev/release-process.md:44]: I may be helpful to indicate that this section is subject to change.

**Review by hackebrot — COMMENTED** [review]
Thank you for your work on this, @Dexterp37! 🚀

File: docs/dev/release-process.md
Comment (Dexterp37) [reply docs/dev/release-process.md:44]: I'd rather not mark content as subject to change for now. I believe the guidelines would be valid even if the context around them would change, but I'd defer this to future PRs to settle. (or really leave it to @ncloudioj )

Comment (ncloudioj) [reply docs/dev/release-process.md:48]: > I'd recommend either leaving this text as it is or extending it in a separate PR as you see fit. I believe there's still value in slightly documenting the process to act on things when prod breaks.

Sounds good to me. I think we will further expand this section with detailed guidance in a follow-up.

**Review by ncloudioj — APPROVED** [review]
@Dexterp37 Thanks for putting this up together!

Let's give it a spin. 🚀

**Comment by Dexterp37** [comment]
> @Dexterp37 Thanks for putting this up together!
> 
> Let's give it a spin. 🚀

Thank you for the review :-)

I'll merge this even if the ci/checks fail, since they are not related to this PR. Note that this still requires a bit of work from Dustin on the Jenkins side to plug things together. He's doing the work today.

**Pull request merged**
